netdev-tc-offloads: Limit the max action number to 16

mishuang2017 opened the pull request:
If there are 32 actions, tfilter_notify() in tc_new_tfilter() will
return -EINVAL because we only allocate 4K buffer which is not big
enough. If we enlarge it to 8K, there is no problem. But customer
doesn't want to change the kernel. So we limit the max action number
to 16 in ovs to workaround this issue.

If tfilter_notify() returns -EINVAL, recvmsg() will return EAGAIN.
In this case, tc_transact() in tc_replace_flower() will return 0.
But the buffer in reply is NULL. So we'll hit the following assert:

./include/openvswitch/ofpbuf.h:189: assertion offset + size <= b->size failed in ofpbuf_at_assert()

Issue: 1917067
Signed-off-by: Chris Mi <chrism@mellanox.com>

roidayan commented:
> If there are 32 actions, tfilter_notify() in tc_new_tfilter() will
> return -EINVAL because we only allocate 4K buffer which is not big

tfilter_notify() is only if err is 0. so how did it return EINVAL?
all EINVAL options are returns and not because of msg size..

we allocate 4k or rtnl api allocates 4k? 
you mean the NLMSG_GOODSIZE right?
if so its not "we" its the kernel net sched api.

I see flower parse to to TCA_ACT_MAX_PRIO which is 32. this is why in ovs we used that as well.


> enough. If we enlarge it to 8K, there is no problem. But customer
> doesn't want to change the kernel.

no need to mention a customer and custom kernel.
we want a fix to support the kernel as is.

> So we limit the max action number
> to 16 in ovs to workaround this issue.
> 
> If tfilter_notify() returns -EINVAL, recvmsg() will return EAGAIN.
> In this case, tc_transact() in tc_replace_flower() will return 0.
> But the buffer in reply is NULL. So we'll hit the following assert:
> 
> ./include/openvswitch/ofpbuf.h:189: assertion offset + size <= b->size failed in ofpbuf_at_assert()
> 
> Issue: 1917067
> Signed-off-by: Chris Mi [chrism@mellanox.com](mailto:chrism@mellanox.com)

roidayan commented on lib/netdev-tc-offloads.c:
```diff
@@ -1283,7 +1283,7 @@ netdev_tc_flow_put(struct netdev *netdev, struct match *match,
 
     NL_ATTR_FOR_EACH(nla, left, actions, actions_len) {
-        if (flower.action_count >= TCA_ACT_MAX_PRIO) {
+        if (flower.action_count >= 16) {
```
we probably need a define for this.

roidayan commented on lib/netdev-tc-offloads.c:
```diff
@@ -1283,7 +1283,7 @@ netdev_tc_flow_put(struct netdev *netdev, struct match *match,
     NL_ATTR_FOR_EACH(nla, left, actions, actions_len) {
-        if (flower.action_count >= TCA_ACT_MAX_PRIO) {
+        if (flower.action_count >= 16) {
             VLOG_DBG_RL(&rl, "Can only support %d actions", flower.action_count);
```
can u do another commit for this series to fix the log. it should print the define and not what the user chose.

roidayan replied:
sorry its not a bug, just looks like this. because action_count changes here. but it will be readable to use the define in the print.

roidayan commented:
a note. TCA_ACT_MAX_PRIO was always 32 and is the definition used in the kernel of how many actions to support and parse from netlink msg.

roidayan commented:
I think we should refactor the msg that the kernel net sched api has a limit of 4K msg size so we should not pass that and limiting the number of actions to 16 is good limitation.


not for the commit msg. do we know we always be good? its also matter of the size of the match. what if the match is big enough or type of action maybe bigger than other action. maybe 16 will also pass the 4k msg size.

mishuang2017 committed:
tc: Limit the max action number to 16

Currently, ovs supports to offload max TCA_ACT_MAX_PRIO(32) actions.
But net sched api has a limit of 4K message size which is not enough
for 32 actions when echo flag is set.

After a lot of testing, we find that 16 actions is a reasonable number.
So in this commit, we introduced a new define to limit the max actions.

Fixes: 0c70132cd288("tc: Make the actions order consistent")
Signed-off-by: Chris Mi <chrism@mellanox.com>
Reviewed-by: Roi Dayan <roid@mellanox.com>
Signed-off-by: Simon Horman <simon.horman@netronome.com>
